core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -289,64 +377,153 @@ Path versionHintFile() {
     return metadataPath(Util.VERSION_HINT_FILENAME);
   }
 
-  private void writeVersionHint(int versionToWrite) {
+  @VisibleForTesting
+  boolean writeVersionHint(FileSystem fs, Integer versionToWrite) throws Exception {
     Path versionHintFile = versionHintFile();
-    FileSystem fs = getFileSystem(versionHintFile, conf);
-
+    Path tempVersionHintFile = metadataPath(UUID.randomUUID() + "-version-hint.temp");
     try {
-      Path tempVersionHintFile = metadataPath(UUID.randomUUID().toString() + "-version-hint.temp");
       writeVersionToPath(fs, tempVersionHintFile, versionToWrite);
-      fs.delete(versionHintFile, false /* recursive delete */);
-      fs.rename(tempVersionHintFile, versionHintFile);
-    } catch (IOException e) {
-      LOG.warn("Failed to update version hint", e);
+      // We can accept that version Hint fails to write, but we can't accept that version Hint
+      // writes the wrong version.
+      if (fs.exists(versionHintFile) || findVersionWithOutVersionHint(fs) != versionToWrite) {
+        throw new AlreadyExistsException(
```

Comment:
Why is this an issue? It just means another process was able to commit n+1 . This shouldn't be an exception

Reply:
Here, as expected, versionHintFile should not appear. If it does, it means that another client is operating. Then we should stop everything.

Reply:
If the versionHint comes up, it means that another client is committing. Even if I don't block it, there's a good chance that fs.rename will get an error. This is because the v+0 version has been committed by another client. 

Also, we need to check to see if a new version has been committed before we commit. If we don't do this, then we have a situation where client B commits version V+1, but has some problems writing to versionHint, and the operation is slow. Client A doesn't have this problem, it's committing V+0. And the versionHint is written successfully. Eventually, versionHint < maxVersion.